##### Importamos librerias

In [2]:
# Tratamiento de datos
import pandas as pd

# Exploración de archivos
import sys
import os

# Funciones personalizadas
sys.path.append(os.path.abspath('../src'))
import sp_eda as se

# Visualizaciones
import matplotlib.pyplot as plt
import seaborn as sns

##### Para mostrar todas las columnas y las filas

In [3]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

##### Cargamos el conjunto de datos desde un archivo CSV

In [4]:
df_bank = pd.read_csv('../data/raw/bank-additional.csv', index_col=0)

## Análisis preliminar df_bank

##### Objetivos de analisis preliminar:
- Analizar la calidad de los datos, detectando valores nulos, duplicados, inconsistencias y posibles outliers.
- Describir estadísticamente las principales variables numéricas y categóricas.

In [5]:
se.analisis_rapido(df_bank)

LAS 5 PRIMERAS COLUMNAS DE DATAFRAME SON:


,age,job,marital,education,default,housing,loan,contact,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,date,latitude,longitude,id_
0,NaN,housemaid,MARRIED,basic.4y,0.0,0.0,0.0,telephone,261,1,999,0,NONEXISTENT,1.1,"93,994","-36,4","4,857",5191,no,2-agosto-2019,41.495,-71.233,089b39d8-e4d0-461b-87d4-814d71e0e079
1,57.0,services,MARRIED,high.school,NaN,0.0,0.0,telephone,149,1,999,0,NONEXISTENT,1.1,"93,994","-36,4",NaN,5191,no,14-septiembre-2016,34.601,-83.923,e9d37224-cb6f-4942-98d7-46672963d097
2,37.0,services,MARRIED,high.school,0.0,1.0,0.0,telephone,226,1,999,0,NONEXISTENT,1.1,"93,994","-36,4","4,857",5191,no,15-febrero-2019,34.939,-94.847,3f9f49b5-e410-4948-bf6e-f9244f04918b
3,40.0,admin.,MARRIED,basic.6y,0.0,0.0,0.0,telephone,151,1,999,0,NONEXISTENT,1.1,"93,994","-36,4",NaN,5191,no,29-noviembre-2015,49.041,-70.308,9991fafb-4447-451a-8be2-b0df6098d13e
4,56.0,services,MARRIED,high.school,0.0,0.0,1.0,telephone,307,1,999,0,NONEXISTENT,1.1,"93,994","-36,4",NaN,5191,no,29-enero-2017,38.033,-104.463,eca60b76-70b6-4077-80ba-bc52e8ebb0eb


INFORMACIÓN BÁSICA DE DATAFRAME:
<class 'pandas.DataFrame'>
Index: 43000 entries, 0 to 27570
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             37880 non-null  float64
 1   job             42655 non-null  str    
 2   marital         42915 non-null  str    
 3   education       41193 non-null  str    
 4   default         34019 non-null  float64
 5   housing         41974 non-null  float64
 6   loan            41974 non-null  float64
 7   contact         43000 non-null  str    
 8   duration        43000 non-null  int64  
 9   campaign        43000 non-null  int64  
 10  pdays           43000 non-null  int64  
 11  previous        43000 non-null  int64  
 12  poutcome        43000 non-null  str    
 13  emp.var.rate    43000 non-null  float64
 14  cons.price.idx  42529 non-null  str    
 15  cons.conf.idx   43000 non-null  str    
 16  euribor3m       33744 non-null  str    
 17  nr.employed   

None

EL NÚMERO DE DUPLICADOS ES: 0
EL NÚMERO DE VALORES NULOS ES:


age               5120
job                345
marital             85
education         1807
default           8981
housing           1026
loan              1026
contact              0
duration             0
campaign             0
pdays                0
previous             0
poutcome             0
emp.var.rate         0
cons.price.idx     471
cons.conf.idx        0
euribor3m         9256
nr.employed          0
y                    0
date               248
latitude             0
longitude            0
id_                  0
dtype: int64

El dataset cuenta con 
43000 filas, donde cada una representa una interacción individual con un cliente durante una campaña de marketing bancario y
23 columnas, que incluyen tanto variables numéricas como variables categóricas.

Columnas del dataset:

●	age: La edad del cliente.

●	job: La ocupación o profesión del cliente.

●	marital: El estado civil del cliente.

●	education: El nivel educativo del cliente.

●	default: Indica si el cliente tiene algún historial de incumplimiento de pagos (1: Sí, 0: No).

●	housing: Indica si el cliente tiene un préstamo hipotecario (1: Sí, 0: No).

●	loan: Indica si el cliente tiene algún otro tipo de préstamo (1: Sí, 0: No).

●	contact: El método de contacto utilizado para comunicarse con el cliente.

●	duration: La duración en segundos de la última interacción con el cliente.

●	campaign: El número de contactos realizados durante esta campaña para este cliente.

●	pdays: Número de días que han pasado desde la última vez que se contactó con el cliente durante esta campaña.

●	previous: Número de veces que se ha contactado con el cliente antes de esta campaña.

●	poutcome: Resultado de la campaña de marketing anterior.

●	emp.var.rate: La tasa de variación del empleo.

●	cons.price.idx: El índice de precios al consumidor.

●	cons.conf.idx: El índice de confianza del consumidor.

●	euribor3m: La tasa de interés de referencia a tres meses.

●	nr.employed: El número de empleados.

●	y: Indica si el cliente ha suscrito un producto o servicio (Sí/No).

●	date: La fecha en la que se realizó la interacción con el cliente.

●	contact_month: Mes en el que se realizó la interacción con el cliente durante la campaña de marketing.***

●	contact_year: Año en el que se realizó la interacción con el cliente durante la campaña de marketing.***

●	id_: Un identificador único para cada registro en el dataset.

*** Las columnas 'latitude' y 'longitude' no estan especificadas en la descripción original del conjunto de datos. En lugar de ellas tenemos especificadas las columnas 'contact_month' y 'contact_year'.


El conjunto de datos **NO** tiene duplicados.

Pero **SI** tiene valores nulos. Vamos a comprobarlos con más detalle:

In [6]:
se.num_nulos(df_bank)

El número de valores nulos:


euribor3m         9256
default           8981
age               5120
education         1807
loan              1026
housing           1026
cons.price.idx     471
job                345
date               248
marital             85
dtype: int64

In [7]:
se.porcent_nulos(df_bank)

,nulos%
euribor3m,21.53
default,20.89
age,11.91
education,4.20
loan,2.39
housing,2.39
cons.price.idx,1.10
job,0.80
date,0.58
marital,0.20


Observaciones sobre los valores nulos:

La variable **euribor3m** presenta el mayor porcentaje de valores nulos (21.53%), lo que puede afectar a los análisis relacionados con indicadores económicos.

**default** y **age** también muestran un porcentaje relevante (20.89% y 11.91%) de datos faltantes, por lo que será necesario evaluar su tratamiento.

Las variables **education**, **loan** y **housing** presentan pocos valores nulos (menos de 5%), con un impacto limitado en el análisis.

El resto de las variables (cons.price.idx, job, date y marital) tienen un porcentaje muy bajo de valores nulos y no representan un problema significativo.

##### Columnas numéricas:

In [8]:
se.descriptivo_num(df_bank)

Variables numéricas son:
 Index(['age', 'default', 'housing', 'loan', 'duration', 'campaign', 'pdays',
       'previous', 'emp.var.rate', 'latitude', 'longitude'],
      dtype='str')

Las estadísticas básicas:


,count,mean,std,min,25%,50%,75%,max
age,37880.0,39.98,10.44,17.0,32.00,38.00,47.00,98.00
default,34019.0,0.00,0.01,0.0,0.00,0.00,0.00,1.00
housing,41974.0,0.54,0.50,0.0,0.00,1.00,1.00,1.00
loan,41974.0,0.16,0.36,0.0,0.00,0.00,0.00,1.00
duration,43000.0,257.74,258.67,0.0,102.00,179.00,319.00,4918.00
campaign,43000.0,2.57,2.77,1.0,1.00,2.00,3.00,56.00
pdays,43000.0,962.33,187.26,0.0,999.00,999.00,999.00,999.00
previous,43000.0,0.17,0.50,0.0,0.00,0.00,0.00,7.00
emp.var.rate,43000.0,0.08,1.57,-3.4,-1.80,1.10,1.40,1.40
latitude,43000.0,36.86,7.23,24.4,30.61,36.76,43.11,49.38


Observación preliminar de los datos numericos:

La variable 'age' muestra que la mayoría de los clientes tiene alrededor de 40 años, aunque existen edades muy bajas y muy altas, lo que indica una población diversa.

Las variables 'default', 'housing' y 'loan' están codificadas numéricamente, representan categorías binarias (1: Sí, 0: No) y se tratarán como variables categóricas en el análisis. Las transformamos en variables de tipo string (yes/no) para que sean más entendibles. 
La mayoría de los clientes no presenta impagos, muchos tienen hipoteca y pocos cuentan con otros préstamos.

La 'duración' de las llamadas presenta mucha variabilidad y algunos valores muy altos, lo que indica la presencia de outliers.

El número de contactos por campaña ('campaign') suele ser bajo, aunque existen algunos casos con muchos intentos (outliers).

La variable 'pdays' se concentra principalmente en el valor 999, lo que indica que muchos clientes no habían sido contactados previamente.

La variable 'previous' muestra que la mayoría de los clientes no tuvo contactos previos antes de la campaña actual.

La variable 'emp.var.rate' presenta variabilidad significativa, lo que puede influir en el comportamiento de los clientes durante las campañas.

Las variables 'latitude' y 'longitude' no estan especificadas en la descripción original del conjunto de datos. En lugar de ellas tenemos especificadas las columnas 'contact_month' y 'contact_year'.
Dado que las variables 'latitude' y 'longitude' no tienen relación con el propósito de nuestro análisis, las eliminamos.


##### Columnas categóricas:

In [11]:
se.descriptivo_cat(df_bank)

Variables categoricas son:
 Index(['job', 'marital', 'education', 'contact', 'poutcome', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y', 'date', 'id_'],
      dtype='str')

Las estadísticas básicas:


,count,unique,top,freq
job,42655,11,admin.,10873
marital,42915,3,MARRIED,25999
education,41193,7,university.degree,12722
contact,43000,2,cellular,27396
poutcome,43000,3,NONEXISTENT,37103
cons.price.idx,42529,26,"93,994",7938
cons.conf.idx,43000,26,"-36,4",8020
euribor3m,33744,309,"4,857",2287
nr.employed,43000,11,"5228,1",16980
y,43000,2,no,38156


La columna *job* tiene 11 valores unicos:
Los 10 valores con más frequencia son:


job
admin.           10873
blue-collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self-employed     1489
housemaid         1123
unemployed        1063
Name: count, dtype: int64

La columna *marital* tiene 3 valores unicos:
Los 10 valores con más frequencia son:


marital
MARRIED     25999
SINGLE      12105
DIVORCED     4811
NaN            85
Name: count, dtype: int64

La columna *education* tiene 7 valores unicos:
Los 10 valores con más frequencia son:


education
university.degree      12722
high.school             9925
basic.9y                6309
professional.course     5477
basic.4y                4356
basic.6y                2386
NaN                     1807
illiterate                18
Name: count, dtype: int64

La columna *contact* tiene 2 valores unicos:
Los 10 valores con más frequencia son:


contact
cellular     27396
telephone    15604
Name: count, dtype: int64

La columna *poutcome* tiene 3 valores unicos:
Los 10 valores con más frequencia son:


poutcome
NONEXISTENT    37103
FAILURE         4461
SUCCESS         1436
Name: count, dtype: int64

La columna *cons.price.idx* tiene 26 valores unicos:
Los 10 valores con más frequencia son:


cons.price.idx
93,994    7938
93,918    6937
92,893    5985
93,444    5349
94,465    4522
93,2      3731
93,075    2552
92,201     803
92,963     742
NaN        471
Name: count, dtype: int64

La columna *cons.conf.idx* tiene 26 valores unicos:
Los 10 valores con más frequencia son:


cons.conf.idx
-36,4    8020
-42,7    7004
-46,2    6057
-36,1    5408
-41,8    4568
-42      3782
-47,1    2581
-31,4     815
-40,8     748
-26,9     477
Name: count, dtype: int64

La columna *euribor3m* tiene 309 valores unicos:
Los 10 valores con más frequencia son:


euribor3m
NaN      9256
4,857    2287
4,962    2124
4,963    2019
4,961    1594
4,964     964
4,856     961
1,405     954
4,864     871
4,965     866
Name: count, dtype: int64

La columna *nr.employed* tiene 11 valores unicos:
Los 10 valores con más frequencia son:


nr.employed
5228,1    16980
5099,1     8937
5191       8020
5195,8     3830
5076,2     1748
5017,5     1133
4991,6      812
5008,7      683
4963,6      665
5023,5      181
Name: count, dtype: int64

La columna *y* tiene 2 valores unicos:
Los 10 valores con más frequencia son:


y
no     38156
yes     4844
Name: count, dtype: int64

La columna *date* tiene 1825 valores unicos:
Los 10 valores con más frequencia son:


date
NaN                   248
28-febrero-2018        95
28-febrero-2016        94
28-febrero-2015        92
28-febrero-2017        89
28-febrero-2019        85
30-abril-2018          58
30-septiembre-2017     56
30-junio-2015          55
30-noviembre-2017      55
Name: count, dtype: int64

La columna *id_* tiene 43000 valores unicos:
Los 10 valores con más frequencia son:


id_
089b39d8-e4d0-461b-87d4-814d71e0e079    1
e9d37224-cb6f-4942-98d7-46672963d097    1
3f9f49b5-e410-4948-bf6e-f9244f04918b    1
9991fafb-4447-451a-8be2-b0df6098d13e    1
eca60b76-70b6-4077-80ba-bc52e8ebb0eb    1
d63ede72-0b6d-45b1-8872-385ac6897f65    1
5e3483e5-236d-437d-8351-541f9d09b9dd    1
87fdc08b-30ae-4dab-803f-561ecdf27ff0    1
87b79988-2be5-419d-88f4-56655852c565    1
ea6b7d04-9271-4c0a-a01f-07795d164aba    1
Name: count, dtype: int64

Observacion preliminar de los datos categóricos:

La variable 'job' incluye varias ocupaciones, siendo admin. la más frecuente.

En 'marital', predominan los clientes casados (married). Cambiaremos los nombres de las categorías a letras minúsculas para estandartizar los datos.

La mayoría de los clientes tiene estudios universitarios, según la variable 'education'.

El contacto con los clientes se realiza principalmente por teléfono móvil (cellular).

En 'poutcome', la mayoría de los valores indican que no hubo campañas previas. Cambiaremos los nombres de las categorías a letras minúsculas para estandartizar los datos.

Las variables 'cons.price.idx', 'cons.conf.idx', 'euribor3m' y 'nr.employed', aunque están representadas como categóricas, corresponden a indicadores económicos con un número limitado de valores únicos, por lo que podrían tratarse como variables numéricas en etapas posteriores del análisis.

La variable 'y' está desbalanceada, con muchos más casos de 'no' que de 'sí'. La variable se presenta como un valor categórico (yes/no). No cambiamos el formato para que sea como otras variables binarias en el conjunto de datos para la unificación de datos. 

La variable 'date' se encuentra como texto, por lo que será necesario convertirla a formato de fecha para permitir un análisis temporal adecuado.

La variable 'id_' contiene valores únicos para cada registro, actuando correctamente como identificador y sin repeticiones. Renombramos la columna 'id_' a 'ID' para mejorar la claridad del nombre.